### PR TITLE
add dynamically-linked musl targets

### PR DESCRIPTION
**Pull request by kaniini**
This pull request adds `$arch-unknown-linux-dynmusl` targets as a generic replacement for Alpine's `$arch-alpine-linux-musl` targets.  The name `dynmusl` was chosen based on discussion in rust-lang/rfcs#2695 and rust-lang/rust#59302.

The platform support entries are based on the current state of the equivalent `$arch-alpine-linux-musl` targets.

Rationale copied from down-thread...

---

The `dynmusl` targets are intended to provide a set of targets intended for use on dynamically-linked musl distros.  They are intended to act the same as the glibc targets.  The `dynmusl` targets are based on Alpine's previous set of targets which did the same thing (and review has found some deficiencies, thanks everyone!).

I believe the `dynmusl` targets provide a compromise that is minimally intrusive to the rust compiler team, while providing musl-based distros like Alpine with a set of targets that meet their requirements.  While not perfect (in an ideal world, the `-musl` targets would not assume static linking and work like glibc), it does solve the problem in a way I believe we can all live with.

These targets are required because Linux distributions generally consider static linking by default to be a policy violation.  It is a policy violation in Alpine, Void and Adelie Linux, which are the top three musl-based distributions.  Though it should also be said that there are staticly-linked musl distributions such as Sabotage, which would continue to be served by the `-musl` targets directly (if they ship Rust at all, I haven't checked).

### CODE REVIEW

File: compiler/rustc_target/src/spec/armv6_unknown_linux_dynmusleabihf.rs [hunk 1]
```diff
@@ -0,0 +1,12 @@
...
+pub fn target() -> Target {
+    let mut base = super::arm_unknown_linux_musleabihf::target();
+
+    base.llvm_target = "armv6-unknown-linux-musleabihf".to_string();
+    base.options.vendor = "alpine".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+    base.options.need_rpath = true;
+
+    base
```
Comment (jyn514) [inline compiler/rustc_target/src/spec/armv6_unknown_linux_dynmusleabihf.rs:12]:
```suggestion
    base
}
```

**Comment by alex** [comment]
Build failure looks relevant.

The docs entries have toolchain entries marked as being built, but I don't see any CI changes. Is there a reason those wouldn't be necessary?

**Review by jyn514 — COMMENTED** [review]
> The docs entries have toolchain entries marked as being built, but I don't see any CI changes. Is there a reason those wouldn't be necessary?

Yeah, I think the checkmarks should be removed from the new entries in `platform-support.md`.

File: compiler/rustc_target/src/spec/armv6_unknown_linux_dynmusleabihf.rs [hunk 2]
```diff
@@ -0,0 +1,12 @@
+use crate::spec::Target;
+
+pub fn target() -> Target {
+    let mut base = super::arm_unknown_linux_musleabihf::target();
+
+    base.llvm_target = "armv6-unknown-linux-musleabihf".to_string();
+    base.options.vendor = "alpine".to_string();
```
Comment (12101111) [inline compiler/rustc_target/src/spec/armv6_unknown_linux_dynmusleabihf.rs:7]: We should not set `vendor` here, as some other Linux distribution also using musl, e.g. Gentoo Linux or OpenWRT

File: src/doc/rustc/src/platform-support.md [hunk 3]
```diff
@@ -158,6 +158,7 @@ target | std | host | notes
 `aarch64-unknown-freebsd` | ✓ | ✓ | ARM64 FreeBSD
 `aarch64-unknown-hermit` | ? |  |
 `aarch64-unknown-linux-gnu_ilp32` | ✓ | ✓ | ARM64 Linux (ILP32 ABI)
+`aarch64-unknown-linux-dynmusl` | ✓ | ✓ | ARM64 Linux (Musl with dynamic linking)
```
Comment (nagisa) [inline src/doc/rustc/src/platform-support.md:161]: I believe you will need to change bootstrap/dockerfiles somewhere under the `src/ci/docker` for builds of libstd/compiler to happen.

cc @Mark-Simulacrum  who knows the setup there better than me.

File: compiler/rustc_target/src/spec/armv7_unknown_linux_dynmusleabihf.rs [hunk 4]
```diff
@@ -0,0 +1,12 @@
...
+pub fn target() -> Target {
+    let mut base = super::armv7_unknown_linux_musleabihf::target();
+
+    base.llvm_target = "armv7-unknown-linux-musleabihf".to_string();
+    base.options.vendor = "alpine".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
+    base.options.need_rpath = true;
+
+    base
```
Comment (ericonr) [inline compiler/rustc_target/src/spec/armv7_unknown_linux_dynmusleabihf.rs:12]: Missing the closing brace here as well.

File: compiler/rustc_target/src/spec/mod.rs [hunk 5]
```diff
@@ -656,17 +662,24 @@ supported_targets! {
     ("armv7-unknown-linux-gnueabihf", armv7_unknown_linux_gnueabihf),
     ("thumbv7neon-unknown-linux-gnueabihf", thumbv7neon_unknown_linux_gnueabihf),
     ("thumbv7neon-unknown-linux-musleabihf", thumbv7neon_unknown_linux_musleabihf),
+    ("armv6-unknown-linux-dynmusleabihf", armv6_unknown_linux_dynmusleabihf),
```
Comment (ericonr) [inline compiler/rustc_target/src/spec/mod.rs:665]: I believe Rust calls this `arm-...`, not `armv6-...`

File: src/doc/rustc/src/platform-support.md (hunk 3 above)
Comment (kaniini) [reply src/doc/rustc/src/platform-support.md]: I cleared the checkmarks for now, we can tackle that in a later MR.

File: compiler/rustc_target/src/spec/armv6_unknown_linux_dynmusleabihf.rs (hunk 2 above)
Comment (kaniini) [reply compiler/rustc_target/src/spec/armv6_unknown_linux_dynmusleabihf.rs]: Yes, this was an error in the original patch.  Fixed now!

File: compiler/rustc_target/src/spec/mod.rs (hunk 5 above)
Comment (kaniini) [reply compiler/rustc_target/src/spec/mod.rs:665]: Seems so, yes.

**Comment by ericonr** [comment]
I have already requested this over IRC, but for completeness sake, i686 dynmusl would complete the target matrix for Void Linux as well.

File: compiler/rustc_target/src/spec/mod.rs (hunk 5 above)
Comment (ericonr) [reply compiler/rustc_target/src/spec/mod.rs:665]: But I might be wrong, there appear to be other armv6 targets...

File: compiler/rustc_target/src/spec/aarch64_unknown_linux_dynmusl.rs [hunk 6]
```diff
@@ -0,0 +1,11 @@
+use crate::spec::Target;
+
+pub fn target() -> Target {
+    let mut base = super::aarch64_unknown_linux_musl::target();
+
+    base.llvm_target = "aarch64-unknown-linux-musl".to_string();
+    base.options.crt_static_default = false;
+    base.options.static_position_independent_executables = true;
```
Comment (12101111) [inline compiler/rustc_target/src/spec/aarch64_unknown_linux_dynmusl.rs:8]: - `base.options.static_position_independent_executables` is default to `true` for linux_base
- `base.pre_link_objects_fallback`, `base.post_link_objects_fallback` and `base.crt_objects_fallback` should not set on dynmusl target ( host toolchain provide those object files ).

**Comment by kaniini** [comment]
@ericonr added in ff1825a :+1:

**Review by 12101111 — COMMENTED** [review]
rustbuild also need to update as `dynmusl` target don't need to ship `rustlib/$target/lib/self-contained/*.o`.

File: compiler/rustc_target/src/spec/aarch64_unknown_linux_dynmusl.rs (hunk 6 above)
Comment (nagisa) [reply compiler/rustc_target/src/spec/aarch64_unknown_linux_dynmusl.rs:8]: Its fine to set the `pie` explicitly here, but the `link_objects` should definitely be removed as that's only necessary for self-contained targets (`musl`)

Comment (nagisa) [reply compiler/rustc_target/src/spec/aarch64_unknown_linux_dynmusl.rs:8]: As for the shipping of the objects, I think you'll just need to add a dash [here](https://github.com/rust-lang/rust/blob/cecdb181ade91c0a5ffab9a148dba68fc7d00ee3/src/bootstrap/compile.rs#L180)

Comment (kaniini) [reply compiler/rustc_target/src/spec/aarch64_unknown_linux_dynmusl.rs]: Should be resolved by c41347cbc62.

File: compiler/rustc_target/src/spec/mod.rs (hunk 5 above)
Comment (kaniini) [reply compiler/rustc_target/src/spec/mod.rs]: On linux it appears to be `arm-`, so I changed the target to that to be aligned.

**Comment by kaniini** [comment]
@petrochenkov

Hi!

I noticed that you responded with :confused: to the beginning of this MR, which concerns me.  So I thought I would explain the rationale behind the `dynmusl` targets for your benefit (and the benefit of others who are not following this issue).

The `dynmusl` targets are intended to provide a set of targets intended for use on dynamically-linked musl distros.  They are intended to act the same as the glibc targets.  The `dynmusl` targets are based on Alpine's previous set of targets which did the same thing (and review has found some deficiencies, thanks everyone!).

I believe the `dynmusl` targets provide a compromise that is minimally intrusive to the rust compiler team, while providing musl-based distros like Alpine with a set of targets that meet their requirements.  While not perfect (in an ideal world, the `-musl` targets would not assume static linking and work like glibc), it does solve the problem in a way I believe we can all live with.

Hopefully this assuages your concerns with the `dynmusl` target and informs everyone at large as to why they exist.

**Comment by ericonr** [comment]
Style: you have `MUSL with dynamic linking` and `(Musl with dynamic linking)` in `platform-support.md`. I don't know which is preferred, but I think it should stick with a single one.

**Comment by workingjubilee** [comment]
The musl website refers to itself consistently in the lowercase, so musl or Musl should be preferred.

**Comment by kaniini** [comment]
I just copied the descriptions from other entries that the `dynmusl` target was derived from.  I can normalize it to `musl`.

**Comment by kaniini** [comment]
I think that this is to a point where we can live with it.  It's not what would exist in an ideal world, but the `-musl` targets came first with a static-linking centric worldview, and breaking them creates problems for those using the targets while solving Musl-based distro problems.  So introducing the `-dynmusl` family of targets seems like a nice compromise we can all live with hopefully.

I'm going to focus on libc crate cleanups for musl in the meantime until everyone who wants to review this has gotten an opportunity to do so.

**Comment by petrochenkov** [comment]
So, I think duplicating the whole set of musl targets to change something that can be changed by a compiler option is an unproportionally heavy hammer for this problem.
That's basically what @mati865 said as far as in 2019 https://github.com/rust-lang/rust/issues/59302#issuecomment-508929674.

The difference between existing musl targets and the new targets introduced in this PR:
- disabling the self-contained mode heuristic (`crt_objects_fallback = None`).
    I'm not sure why this is done, `fn crt_objects_fallback` in `link.rs` (which determines the effective fallback based on target defaults and current options) already disables the fallback if musl is linked dynamically. If necessary the heuristic can be tweaked further (cc https://github.com/rust-lang/rust/pull/71769#issuecomment-626330237).
- switching default libc linking to dynamic, which is supposed to be done with `-C target-feature=-crt-static`.
    The main question is - why is `-C target-feature=+crt-static` not enough and why do we need a target?
    Should we introduce new targets for *any* compiler options? I don't think so. Then why is `-C target-featue` special? Do we need a separate set of targets for statically linked glibc, for example?
    If cargo makes passing such options inconvenient, then it should be pushed to make compiler option support more convenient and fine-grained instead.

**Comment by kaniini** [comment]
> So, I think duplicating the whole set of musl targets to change something that can be changed by a compiler option is an unproportionally heavy hammer for this problem.

I agree that this approach is *extremely* heavy-handed, but musl-based distributions need a different default behavior than what the `-musl` targets presently do.

Consensus so far has pointed to introducing `-dynmusl` for this purpose.  Right now, every single musl-based distribution has patches like these, or they just patch the default `-musl` targets, leading to inconsistent behavior across musl-based distros.

We would like to drop these patches and find a solution that works for us.  Do you have a proposal for an alternative solution that will work for our use case?

To make it clear:

What musl-based distros want: `rustc` generates dynamically-linked binaries *by default*.
What `rustc` presently does without patching on musl-based distros: generates staticly-linked binaries by default.

We do not want to carry these patches anymore, and would like to come to a permanent solution that works well for musl-based distros *and* the rust community at large.

I'm open to alternative suggestions that allow for changing the default behavior other than introducing new targets, but that's what was proposed every step of the way so far.

**Comment by nagisa** [comment]
>  The main question is - why is -C target-feature=-crt-static not enough and why do we need a separate target?

Configuring the exact target behaviour via flags is "good enough" when all you're concerned about is building code targetting a specific set of constraints.

Flags don't work great however, if you want to run rustc on a "dynmusl" host (such as what Alpine and Void are) and get the same kind of seamless experience as you'd get when running on the more traditional targets (such as linux-gnu or windows), there's no real choice other than adding a target to rustc _somehow_. Until now these platforms would patch the targets into their own builds. But this has started causing some [issues for these platforms](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/bootstrapping.20on.20s390x.2Fmips.20for.20musl) recently.

IOW, these targets describe not a difference in a single flag, but a different family of platforms. In that particular respects `dynmusl` targets are actually more conventional (and closely match the existing `-gnu` and `-msvc` targets) than the `musl` targets we currently have.

**Comment by kaniini** [comment]
> IOW, these targets describe not a difference in a single flag, but a different family of platforms. In that particular respects `dynmusl` targets are actually more conventional (and closely match the existing `-gnu` and `-msvc` targets) than the `musl` targets we currently have.

Precisely this.  The `-musl` targets in rust are optimized for creating static-binaries.  The intention of the `-dynmusl` targets is to eventually diverge from the `-musl` targets to provide an optimized experience for use on musl-based distributions.

Regarding https://github.com/rust-lang/rust/issues/59302#issuecomment-508929674, I agree that is a better approach, but we do not have the resources to cheer-lead that at present.  Implementing the `-dynmusl` family of targets solves an immediate problem for rust-based distributions and would allow us to migrate to a better solution later.

**Comment by workingjubilee** [comment]
If the problem is that it's mostly enabled but the build system does not flag for dynlinking by default, is it possible to handle this at the build system level, then? Possibly using one of the existing configuration bits on the system we're reading in anyways when we determine what our RUSTFLAGS are and such.

Mmm, but the concern merely saying that raises in the back of my brain would be the problem of configuration management over time, possibly turning into buggy garbage pretty quickly.

**Comment by kaniini** [comment]
I don't care if the build system (presumably cargo) manages it, as long as this behavior is transparent.  We want the tools to behave transparently at runtime: just like GCC gives you a dynamically linked binary on Alpine, we want rustc to as well, without having to fiddle with anything, and without having to patch anything.

**Comment by mati865** [comment]
Thanks for the CC.
I was unable to find enough time and energy to chase my idea so I won't stop anyone from fixing/hacking it differently.

**Comment by kaniini** [comment]
> disabling the self-contained mode heuristic (crt_objects_fallback = None).
> I'm not sure why this is done, fn crt_objects_fallback in link.rs (which determines the effective fallback based on target defaults and current options) already disables the fallback if musl is linked dynamically.

This was requested in review by @12101111.  If it is not needed, we can always drop it.  In Alpine, we have not been doing this and it does seem fine.

>     switching default libc linking to dynamic, which is supposed to be done with -C target-feature=-crt-static.
>    The main question is - why is -C target-feature=-crt-static not enough and why do we need a separate target?
>    Should we introduce new targets for any compiler options? I don't think so. Then why is -C target-featue special? Do we also need a separate set of targets for statically linked glibc, for example?
>    If cargo makes passing such options inconvenient, then it should be pushed to make compiler option support more convenient and fine-grained instead.

The problem is basically one of *equity*.  The original `-musl` targets have behaviour that is not suitable for musl-based distributions, which were not consulted when the `-musl` targets were created.  Allowing musl-based distributions to have their own family of targets enables us to appropriately define policy for how rust should behave on musl-based distributions.

I am open to other approaches but are concerned with the fragility of those approaches.  For example, in the past, we have had people edit the GCC specs file on their system and wind up breaking their compiler.  So I am hesitant to go with a config-based approach here.  We also do not want to tell people that they have to use custom `RUSTFLAGS` in order to get the same behavior as the rest of their tools.  And we don't think it is fair that we have to patch rust to get the right behavior for musl-based distributions when glibc distributions get the right behavior out of the box.

All in all, I still believe this approach is the most flexible (allowing us to expand on policy changes in the future) while imposing minimal burden on the rust team.  It is a compromise -- one that most people will not be thrilled about, but one that everyone can live with and get an acceptable outcome.

**Comment by nagisa** [comment]
@rfcbot fcp merge

This PR adds a large number of targets describing a family of platforms/host based on musl, much like the `-gnu` and `-msvc` targets do. 

Given the motivations expressed in this PR as well as in this [zulip thread](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/bootstrapping.20on.20s390x.2Fmips.20for.20musl) I'm of an opinion that this should be merged. There is a concern raised by [petrochenkov](https://github.com/rust-lang/rust/pull/82556#issuecomment-786945801) which I believe to be addressed to a good degree. As thus I'm kicking off the FCP process to draw attention of other members of the T-compiler, as well as give somewhat more defined timeline for concerns to be raised/discussed/resolved.

**Comment by petrochenkov** [comment]
Ok, I see, on musl hosts you want to pass `-C target-feature=-crt-static` almost always, and that includes cases like compiling "Hello world" when you want things to "just work" by default.
So `crt_static_default` does need to be effectively `false` on musl hosts.

I don't see why target defaults cannot depend on the host though.

>musl-based distribution ... just patch the default `-musl` targets

This is what I've seen in one of the previous threads, and I think it's pretty reasonable to move such a patch to upstream rustc.
It it trivially implementable and should look approximately like this - https://github.com/petrochenkov/rust/commit/dd9692c6778d37ad01e1de5ecb9fcb8054557fe1.

We already have a precedent for things like this on Apple targets where rustc fetches some target details from the host environment, see https://github.com/rust-lang/rust/blob/master/compiler/rustc_target/src/spec/apple_base.rs.

Besides musl targets we also have several targets that use static linking by default - redox, vxworks and wasi.
If they end up being used as rustc hosts, I wouldn't want to see them duplicated as well, while switching linking to dynamic depending on host would work for them as well.

**Comment by kaniini** [comment]
I support @petrochenkov's proposal to make the `-musl` targets smarter than duplicating.  That solves our problem without duplicating the targets.

**Comment by kaniini** [comment]
With that said in Zulip, @nagisa brings up concerns that the behaviour is inconsistent.

I think the right thing here is to fix the `-musl` targets and make cross-compilers specify they want static linking.  This is also what @richfelker (the musl author) has asked for in #alpine-devel on IRC.

The reality is that the hack of defining `-musl` targets to be static by default was never a correct thing to do, despite it mostly working and any conversation with anybody doing work in the musl ecosystem would have verified this.  While static linking under musl is a mostly pleasant experience, there are still caveats.

**Comment by richfelker** [comment]
Yes, as main author and maintainer of musl, I would very much prefer that Rust not be treating it oddly here regardless of whether the host is musl-based or you're cross-compiling for a musl-based target. While I'm very much a proponent of static linking and distributing static binaries that do not depend on anything about the library ecosystem of the system they'll run on, musl-based hosts with dynamic linking are also an important cross-compiling target in themselves, especially in the embedded space. I'm not sure where the original Rust idea that "musl is static linking" came from, but the choice of musl and the choice to static link should be completely independent.

**Comment by nagisa** [comment]
Replicating my thoughts from Zulip:

I don't have particularly strong opinions on which approach is taken as long as we end up supporting the dynamic musl platforms properly. That said, I think there are a number of downsides to making the `musl` target switch between `+crt-static` and `-crt-static` depending on runtime environment.

First and foremost it will confuse people on dynmusl platforms who read blog posts suggesting that `--target=*-musl` is all you need to do get fully statically linked binaries. They for some reason would keep getting `-crt-static` outputs. In a similar vein, crossing to `musl` platforms from e.g. `-gnu` would also behave quite weirdly, compared to the other targets.

Another question is what does `*-unknown-linux-musl` mean exactly in that kind of setup when we're talking about toolchains? Even for the targets we ship right now it is quite confusing – the compiler we ship for the target is `-crt-static`, despite every material out here suggesting the target is `+crt-static` all around.

I think macOS is a pretty poor model to follow. You cannot feasibly cross to those targets in the first place, so a large number of concerns that would be relevant to other targets are not relevant for Apple. Apple also have a number of their own conventions (e.g. `_DEPLOYMENT_TARGET`) baked into their native toolchains – people expect those to work in a specific way that makes dynamic targets for Apple stuff somewhat unavoidable.

**Comment by kaniini** [comment]
My suggestion would be:

- Just fix the `-musl` targets to behave like conventional targets, since musl is most widely used and deployed in a conventional setting.

- Add a warning when building with the `-musl` targets that the static-linking default was changed and provide a pointer to how to enable it.

- Musl distros can patch out the warning for now until it is removed.

**Comment by workingjubilee** [comment]
Any solution regardless will have to be heavily "marketed", any which way. The concern regarding invaliding existing advisory material is a real one, but if we use "dynmusl" targets, then people who are expecting to use Rust on the dynmusl targets will have to be told that they need to use dynmusl, and they'll read blog posts which suggest other behavior and get confused too, and compile a static binary when they wanted a dynamic one, or vice versa. And the current situation is, as nagisa noted, heavily contradictory on the face of it when shipping Rust for musl targets.

There's no real replacement for good SEO.

**Comment by kaniini** [comment]
My understanding is that when you download a `-musl` toolchain with rustup, you don't actually get a musl toolchain at all, but instead something which has musl embedded into `libstd`.

So perhaps there is value to keeping that target around as `-embmusl` (embedded musl), which would be descriptive of what you're actually getting.  Then `-musl` can live a a conventional target, which is aligned with what the world at large would generally expect.

**Comment by petrochenkov** [comment]
One more data point - musl targets are currently not always static by default:
- static-by-default: arm/aarch64, i686/x86_64, mips (some targets), powerpc, riscv, s390x
- dynamic-by-default: mips (some targets), hexagon

I don't know what it the story behind these differences.

**Comment by petrochenkov** [comment]
Can we switch everything to dynamic on nightly and see what breaks in practice? :)

I suspect it will be the `x86_64-unknown-linux-gnu` -> `x86_64-unknown-linux-musl` cross-compilation scenario first of all.
We can then special case it for backward compatibility.

Also, I think we can just break Tier 3 targets freely, and change the behavior to whatever we want.
I'm not sure about Tier 2 targets though, probably not.
(Note that all musl targets are Tier 2 and Tier 3, there's not Tier 1 support.)

**Comment by ericonr** [comment]
> I suspect it will be the x86_64-unknown-linux-gnu -> x86_64-unknown-linux-musl cross-compilation scenario first of all.
We can then special case it for backward compatibility.

Do you have an idea what it would look like?  Cross compilation from glibc to musl *is* a use case Void Linux supports (though not what we currently use in builders), and special casing musl there would mean we'd still need to patch Rust. Unless you meant specifically `x86_64` to `x86_64`, which I believe should be mostly harmless.

**Comment by petrochenkov** [comment]
@ericonr 
After some thought, it wasn't a good idea, 
More fine-grained per-target and per-host configuration would only mean more hackery for the musl target and less predictability.

**Comment by petrochenkov** [comment]
Do I understand correctly that people working on musl or with musl would prefer a standard glibc-like behavior (without static linking by default, and without self-contained linking heuristics) if musl support was done from scratch?
Also do I understand correctly that long term you'd like to migrate the existing musl targets to that behavior, if not for breakage concerns?

In that case I must agree that adding new targets is the most predictable and hack-free solution.
If we are doing that, we should be ready to follow this path consistently though, rather than just to add duplicate targets.
- Choose some target name conveying the "musl done right"/"musl behaving in standard way"/"better musl" meaning rather than "musl with dynamic linking". (Let's call it musl2 for now because 2 is better than 1, you know.)
- Immediately migrate tier 3 targets to musl2 instead of duplicating them, because we can break them.
- Immediately migrate targets with dynamic-by-default musl (https://github.com/rust-lang/rust/pull/82556#issuecomment-787429950) to musl2.
- Keep remaining tier2 musl targets as is for compatibility, update all docs like https://doc.rust-lang.org/edition-guide/rust-2018/platform-and-target-support/musl-support-for-fully-static-binaries.html

What do you think about this plan?

**Comment by richfelker** [comment]
I don't like the naming "musl2" or having plain "musl" naming continue to be some weird static-by-default thing. Please just fix it. At worst some build recipes break. Nothing breaks in deployment of existing applications.

**Comment by petrochenkov** [comment]
>At worst some build recipes break.

Yeah, and that's the whole problem, we've been advertising this behavior as stable for years.

**Comment by richfelker** [comment]
I'm in favor of changing it on nightly and seeing what breaks.

**Comment by estebank** [comment]
@petrochenkov we _could_ change the default behavior and when detecting we are in a non-musl platform (the opposite check you linked to earlier) emit an unconditional warning letting you know that if you meant to build a statically compiled musl binary, to be explicit by passing `+crt-static` or to silence the warning by explicitly passing `-crt-static`. The run to see what breaks would still be needed and the communication effort will be significant. We wouldn't be able to remove the warning for _years_ due to the existing, out of date documentation out there that is not under our control.

**Comment by kaniini** [comment]
I think what @estebank proposes makes the most sense.  What would be involved in adding a warning for this?

**Comment by michaelforney** [comment]
Note also that some build recipes are *already* broken with the current static-by-default behavior.

> The main question is - why is `-C target-feature=-crt-static` not enough and why do we need a separate target?

Here's an example: firefox requires its build scripts to be dynamically linked, but it uses `cargo --target=...` which makes it impossible set flags for the build scripts, thus it is currently impossible to build firefox with *-unknown-linux-musl rust toolchains.

I think everyone in the musl community would really just like to see this fixed using the standard toolchain names. It has been a huge barrier for people trying to use rust on musl-based systems for years now.

**Comment by petrochenkov** [comment]
@michaelforney 
>requires its build scripts to be dynamically linked, but it uses `cargo --target=...` which makes it impossible set flags for the build scripts

This is exactly the kind of issue that I meant in

>If cargo makes passing such options inconvenient, then it should be pushed to make compiler option support more convenient and fine-grained instead.

in https://github.com/rust-lang/rust/pull/82556#issuecomment-786945801 (regardless of the musl specifics).

**Comment by kaniini** [comment]
I agree that it should be easier to override target behaviour but in this case the default behaviour is completely broken for musl-based distributions, which should be the primary use case for a `-musl` target, like glibc-based distros are the primary use case for the `-gnu` targets.

**Comment by workingjubilee** [comment]
Wait, does setting `RUSTFLAGS="-Ctarget-feature=-crt-static"` with `cargo build --target x86_64-unknown-linux-musl` override RUSTFLAGS? I just want to make sure if I am understanding the existing behavior correctly.

**Comment by bjorn3** [comment]
If you pass `--target`, cargo ignores `RUSTFLAGS` for crates build for the host like build scripts and proc macros.

**Comment by richfelker** [comment]
> I agree that it should be easier to override target behaviour but in this case the default behaviour is completely broken for musl-based distributions, which should be the primary use case for a `-musl` target, like glibc-based distros are the primary use case for the `-gnu` targets.

Yes. This is the case where the application being compiled and build script being used to compile it have no reason to know anything about musl or that the target is musl-based, and they can't be expected to do anything to override the wrong default of static linking that might break what they're trying to do.

If on the other hand youre trying to make static binaries using musl to distribute and run on systems where they're foreign, you're in the case where you're not just using a target fed to you as whatever the host default is, but working with a specific known system. In that case you can reasonably be expected to take care of your own target options.

**Comment by nagisa** [comment]
@rfcbot fcp cancel

Given the discussion is ongoing and appears to have leaned towards an alternate approach, I'm going to cancel the FCP proposing the merge of this PR as it is written currently.

**Comment by estebank** [comment]
@kaniini just want to let you know that fcp cancellation is about the specifics of this PR's approach, given that there are multiple possible courses of action and we need to reach an agreement on which one to pursue. We still want to see this fixed in a way that everybody involved is happy with!

**Comment by apiraino** [comment]
Given [this comment](https://github.com/rust-lang/rust/pull/82556#issuecomment-790743261) I'm removing also the "waiting on team" label for now.

@rustbot label -S-waiting-on-team

**Comment by petrochenkov** [comment]
This is still waiting-on-team, the issue was nominated for discussion, but not discussed yet. [Summary](https://rust-lang.zulipchat.com/#narrow/stream/227806-t-compiler.2Fwg-prioritization/topic/2021-03-03.20meeting.20agenda/near/228681359).

**Comment by apiraino** [comment]
@petrochenkov ah sorry then I have misunderstood the discussion from last week [T-compiler meeting](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings/topic/.5Bweekly.20meeting.5D.202021-03-04.20.2354818) about this. Thanks for fixing!

**Comment by wesleywiser** [comment]
Hi everyone! We have a compiler team design meeting scheduled to discuss what to do to resolve this issue on March 26 10am EDT (14:00 UTC). Meetings are held in the [`#t-compiler/meetings` Zulip stream](https://rust-lang.zulipchat.com/#narrow/stream/238009-t-compiler.2Fmeetings).

**Comment by pnkfelix** [comment]
The compiler team had a [design team meeting](https://zulip-archive.rust-lang.org/238009tcompilermeetings/10080designmeeting20210326musllinkingcompilerteam416.html)  have an MCP with an alternative solution to this problem: https://github.com/rust-lang/compiler-team/issues/422

**Comment by kaniini** [comment]
I think we can close this PR and focus on the other solution.

**Pull request closed without merging**